packages/astro/src/actions/runtime/virtual/server.ts

```diff
@@ -0,0 +1,141 @@
+import {
+	ActionError,
+	ActionInputError,
+	callSafely,
+	type ErrorInferenceObject,
+	type SafeResult,
+} from './shared.js';
+
+export * from './shared.js';
+
+export { z } from 'zod';
+
+export { getApiContext } from '../store.js';
+
+export function defineAction<
+	TOutput,
+	TAccept extends 'form' | 'json' = 'json',
+	TInputSchema extends TAccept extends 'form'
+		? z.AnyZodObject | z.ZodType<FormData>
+		: z.ZodType = TAccept extends 'form' ? z.ZodType<FormData> : never,
+>({
+	accept,
+	input: inputSchema,
+	handler,
+}: {
+	input?: TInputSchema;
+	accept?: TAccept;
+	handler: (input: z.infer<TInputSchema>) => MaybePromise<TOutput>;
+}): ((
+	input: TAccept extends 'form' ? FormData : z.input<TInputSchema>
+) => Promise<Awaited<TOutput>>) & {
+	safe: (
+		input: TAccept extends 'form' ? FormData : z.input<TInputSchema>
+	) => Promise<
+		SafeResult<
+			z.infer<TInputSchema> extends ErrorInferenceObject
+				? z.infer<TInputSchema>
+				: ErrorInferenceObject,
+			Awaited<TOutput>
+		>
+	>;
+} {
+	const serverHandler =
+		accept === 'form'
+			? getFormServerHandler(handler, inputSchema)
+			: getJsonServerHandler(handler, inputSchema);
+
+	(serverHandler as any).safe = async (
+		unparsedInput: unknown
+	): Promise<SafeResult<TInputSchema, Awaited<TOutput>>> => {
+		return callSafely(() => serverHandler(unparsedInput));
+	};
+	return serverHandler as any;
+}
+
+function getFormServerHandler<TOutput, TInputSchema extends z.AnyZodObject | z.ZodType<FormData>>(
+	handler: (input: z.infer<TInputSchema>) => MaybePromise<TOutput>,
+	inputSchema?: TInputSchema
+) {
+	return async (unparsedInput: unknown): Promise<Awaited<TOutput>> => {
+		if (!(unparsedInput instanceof FormData)) {
+			throw new ActionError({
+				code: 'UNSUPPORTED_MEDIA_TYPE',
+				message: 'This action only accepts FormData.',
+			});
+		}
+
+		if (!(inputSchema instanceof z.ZodObject)) return await handler(unparsedInput);
+
+		const parsed = await inputSchema.safeParseAsync(upgradeFormData(unparsedInput, inputSchema));
+		if (!parsed.success) {
+			throw new ActionInputError(parsed.error.issues);
+		}
+		return await handler(parsed.data);
+	};
+}
+
+function getJsonServerHandler<TOutput, TInputSchema extends z.ZodType<unknown>>(
+	handler: (input: z.infer<TInputSchema>) => MaybePromise<TOutput>,
+	inputSchema?: TInputSchema
+) {
+	return async (unparsedInput: unknown): Promise<Awaited<TOutput>> => {
+		const context = getApiContext();
+		if (context.request.headers.get('content-type') !== 'application/json') {
+			throw new ActionError({
+				code: 'UNSUPPORTED_MEDIA_TYPE',
+				message: 'This action only accepts JSON.',
+			});
+		}
+
+		if (!inputSchema) return await handler(unparsedInput);
+		const parsed = await inputSchema.safeParseAsync(unparsedInput);
+		if (!parsed.success) {
+			throw new ActionInputError(parsed.error.issues);
+		}
+		return await handler(parsed.data);
+	};
+}
+
+export function upgradeFormData<T extends z.AnyZodObject>(
```

Comment:
Let's comment on this function. What does upgrading a form mean?

Reply:
Fair. I'll try a clearer function name, or a comment if I can't come up with one.